openff/toolkit/topology/molecule.py

```diff
@@ -5911,21 +5912,23 @@ def perceive_hierarchy(self):
 
         self.hierarchy_elements = list()
         # Determine which atoms should get added to which HierarchyElements
-        hier_eles_to_add = defaultdict(list)
+        hier_eles_to_add: defaultdict[tuple[Union[int, str]], list["Atom"]] = (
+            defaultdict(list)
+        )
         for atom in self.parent.atoms:
-            atom_key = list()
+            _atom_key = list()
             for field_key in self.uniqueness_criteria:
                 if field_key in atom.metadata:
-                    atom_key.append(atom.metadata[field_key])
+                    _atom_key.append(atom.metadata[field_key])
                 else:
-                    atom_key.append("None")
+                    _atom_key.append("None")
 
-            hier_eles_to_add[tuple(atom_key)].append(atom)
+            hier_eles_to_add[tuple(_atom_key)].append(atom)
```

Comment:
In this case it's probably because re-using `atom_key` would have the same-named variable with different types. Just copying from my local version:

```python
        for atom in self.parent.atoms:
            _atom_key = list()
            for field_key in self.uniqueness_criteria:
                if field_key in atom.metadata:
                    _atom_key.append(atom.metadata[field_key])
                else:
                    _atom_key.append("None")

            hier_eles_to_add[tuple(_atom_key)].append(atom)

        # Create the actual HierarchyElements
        for atom_key, atoms_to_add in hier_eles_to_add.items():
            atom_indices = [p.molecule_atom_index for p in atoms_to_add]
            self.add_hierarchy_element(identifier=atom_key, atom_indices=atom_indices)
```

It's built up as a list and needs to get cast to tuple when it's inserted, but when it's accessed later it's already a tuple. Probably this confused me at some point and the underscore was simpler than having `atom_keys_im_working_on_building_up` and `atom_keys_we_are_looking_at_now`. I sometimes like prepending variable names when they're not returned, but I don't think that's the case here